diff --git a/src/BitfieldMan.cc b/src/BitfieldMan.cc
index 8d1d189b..8b32becd 100644
--- a/src/BitfieldMan.cc
+++ b/src/BitfieldMan.cc
@@ -171,17 +171,16 @@ bool BitfieldMan::hasMissingPiece(const unsigned char* peerBitfield,
   return retval;
 }
 
-bool BitfieldMan::getFirstMissingUnusedIndex(size_t& index) const
+std::optional<size_t> BitfieldMan::getFirstMissingUnusedIndex() const
 {
   if (filterEnabled_) {
     return bitfield::getFirstSetBitIndex(
-        index,
         ~array(bitfield_) & ~array(useBitfield_) & array(filterBitfield_),
         blocks_);
   }
   else {
     return bitfield::getFirstSetBitIndex(
-        index, ~array(bitfield_) & ~array(useBitfield_), blocks_);
+        ~array(bitfield_) & ~array(useBitfield_), blocks_);
   }
 }
 
@@ -201,14 +200,14 @@ size_t BitfieldMan::getFirstNMissingUnusedIndex(std::vector<size_t>& out,
   }
 }
 
-bool BitfieldMan::getFirstMissingIndex(size_t& index) const
+std::optional<size_t> BitfieldMan::getFirstMissingIndex() const
 {
   if (filterEnabled_) {
     return bitfield::getFirstSetBitIndex(
-        index, ~array(bitfield_) & array(filterBitfield_), blocks_);
+        ~array(bitfield_) & array(filterBitfield_), blocks_);
   }
   else {
-    return bitfield::getFirstSetBitIndex(index, ~array(bitfield_), blocks_);
+    return bitfield::getFirstSetBitIndex(~array(bitfield_), blocks_);
   }
 }
 
@@ -241,10 +240,10 @@ size_t getEndIndex(size_t index, const Array& bitfield, size_t blocks)
 
 namespace {
 template <typename Array>
-bool getSparseMissingUnusedIndex(size_t& index, int32_t minSplitSize,
-                                 const Array& bitfield,
-                                 const unsigned char* useBitfield,
-                                 int32_t blockLength, size_t blocks)
+std::optional<size_t>
+getSparseMissingUnusedIndex(int32_t minSplitSize, const Array& bitfield,
+                            const unsigned char* useBitfield,
+                            int32_t blockLength, size_t blocks)
 {
   BitfieldMan::Range maxRange;
   BitfieldMan::Range currentRange;
@@ -277,8 +276,7 @@ bool getSparseMissingUnusedIndex(size_t& index, int32_t minSplitSize,
   }
   if (maxRange.getSize()) {
     if (maxRange.startIndex == 0) {
-      index = 0;
-      return true;
+      return size_t{0};
     }
     else {
       if ((!bitfield::test(useBitfield, blocks, maxRange.startIndex - 1) &&
@@ -286,34 +284,33 @@ bool getSparseMissingUnusedIndex(size_t& index, int32_t minSplitSize,
           (static_cast<int64_t>(maxRange.endIndex - maxRange.startIndex) *
                blockLength >=
            minSplitSize)) {
-        index = maxRange.startIndex;
-        return true;
+        return maxRange.startIndex;
       }
       else {
-        return false;
+        return std::nullopt;
       }
     }
   }
   else {
-    return false;
+    return std::nullopt;
   }
 }
 } // namespace
 
-bool BitfieldMan::getSparseMissingUnusedIndex(
-    size_t& index, int32_t minSplitSize, const unsigned char* ignoreBitfield,
+std::optional<size_t> BitfieldMan::getSparseMissingUnusedIndex(
+    int32_t minSplitSize, const unsigned char* ignoreBitfield,
     size_t ignoreBitfieldLength) const
 {
   if (filterEnabled_) {
     return aria2::getSparseMissingUnusedIndex(
-        index, minSplitSize,
+        minSplitSize,
         array(ignoreBitfield) | ~array(filterBitfield_) | array(bitfield_) |
             array(useBitfield_),
         useBitfield_, blockLength_, blocks_);
   }
   else {
     return aria2::getSparseMissingUnusedIndex(
-        index, minSplitSize,
+        minSplitSize,
         array(ignoreBitfield) | array(bitfield_) | array(useBitfield_),
         useBitfield_, blockLength_, blocks_);
   }
@@ -321,16 +318,16 @@ bool BitfieldMan::getSparseMissingUnusedIndex(
 
 namespace {
 template <typename Array>
-bool getGeomMissingUnusedIndex(size_t& index, int32_t minSplitSize,
-                               const Array& bitfield,
-                               const unsigned char* useBitfield,
-                               int32_t blockLength, size_t blocks, double base,
-                               size_t offsetIndex)
+std::optional<size_t>
+getGeomMissingUnusedIndex(int32_t minSplitSize, const Array& bitfield,
+                          const unsigned char* useBitfield,
+                          int32_t blockLength, size_t blocks, double base,
+                          size_t offsetIndex)
 {
   double start = 0;
   double end = 1;
   while (start + offsetIndex < blocks) {
-    index = blocks;
+    size_t found = blocks;
     for (size_t i = start + offsetIndex,
                 eoi = std::min(blocks, static_cast<size_t>(end + offsetIndex));
          i < eoi; ++i) {
@@ -338,39 +335,37 @@ bool getGeomMissingUnusedIndex(size_t& index, int32_t minSplitSize,
         break;
       }
       else if (!bitfield::test(bitfield, blocks, i)) {
-        index = i;
+        found = i;
         break;
       }
     }
-    if (index < blocks) {
-      return true;
+    if (found < blocks) {
+      return found;
     }
     else {
       start = end;
       end *= base;
     }
   }
-  return getSparseMissingUnusedIndex(index, minSplitSize, bitfield, useBitfield,
+  return getSparseMissingUnusedIndex(minSplitSize, bitfield, useBitfield,
                                      blockLength, blocks);
 }
 } // namespace
 
-bool BitfieldMan::getGeomMissingUnusedIndex(size_t& index, int32_t minSplitSize,
-                                            const unsigned char* ignoreBitfield,
-                                            size_t ignoreBitfieldLength,
-                                            double base,
-                                            size_t offsetIndex) const
+std::optional<size_t> BitfieldMan::getGeomMissingUnusedIndex(
+    int32_t minSplitSize, const unsigned char* ignoreBitfield,
+    size_t ignoreBitfieldLength, double base, size_t offsetIndex) const
 {
   if (filterEnabled_) {
     return aria2::getGeomMissingUnusedIndex(
-        index, minSplitSize,
+        minSplitSize,
         array(ignoreBitfield) | ~array(filterBitfield_) | array(bitfield_) |
             array(useBitfield_),
         useBitfield_, blockLength_, blocks_, base, offsetIndex);
   }
   else {
     return aria2::getGeomMissingUnusedIndex(
-        index, minSplitSize,
+        minSplitSize,
         array(ignoreBitfield) | array(bitfield_) | array(useBitfield_),
         useBitfield_, blockLength_, blocks_, base, offsetIndex);
   }
@@ -378,17 +373,16 @@ bool BitfieldMan::getGeomMissingUnusedIndex(size_t& index, int32_t minSplitSize,
 
 namespace {
 template <typename Array>
-bool getInorderMissingUnusedIndex(size_t& index, size_t startIndex,
-                                  size_t lastIndex, int32_t minSplitSize,
-                                  const Array& bitfield,
-                                  const unsigned char* useBitfield,
-                                  int32_t blockLength, size_t blocks)
+std::optional<size_t>
+getInorderMissingUnusedIndex(size_t startIndex, size_t lastIndex,
+                             int32_t minSplitSize, const Array& bitfield,
+                             const unsigned char* useBitfield,
+                             int32_t blockLength, size_t blocks)
 {
   // We always return first piece if it is available.
   if (!bitfield::test(bitfield, blocks, startIndex) &&
       !bitfield::test(useBitfield, blocks, startIndex)) {
-    index = startIndex;
-    return true;
+    return startIndex;
   }
   for (size_t i = startIndex + 1; i < lastIndex;) {
     if (!bitfield::test(bitfield, blocks, i) &&
@@ -397,8 +391,7 @@ bool getInorderMissingUnusedIndex(size_t& index, size_t startIndex,
       // from this index.
       if (!bitfield::test(useBitfield, blocks, i - 1) &&
           bitfield::test(bitfield, blocks, i - 1)) {
-        index = i;
-        return true;
+        return i;
       }
       // Check free space of minSplitSize.  When checking this, we use
       // blocks instead of lastIndex.
@@ -409,8 +402,7 @@ bool getInorderMissingUnusedIndex(size_t& index, size_t startIndex,
           break;
         }
         if (static_cast<int64_t>(j - i + 1) * blockLength >= minSplitSize) {
-          index = j;
-          return true;
+          return j;
         }
       }
       i = j + 1;
@@ -419,44 +411,44 @@ bool getInorderMissingUnusedIndex(size_t& index, size_t startIndex,
       ++i;
     }
   }
-  return false;
+  return std::nullopt;
 }
 } // namespace
 
-bool BitfieldMan::getInorderMissingUnusedIndex(
-    size_t& index, int32_t minSplitSize, const unsigned char* ignoreBitfield,
+std::optional<size_t> BitfieldMan::getInorderMissingUnusedIndex(
+    int32_t minSplitSize, const unsigned char* ignoreBitfield,
     size_t ignoreBitfieldLength) const
 {
   if (filterEnabled_) {
     return aria2::getInorderMissingUnusedIndex(
-        index, 0, blocks_, minSplitSize,
+        0, blocks_, minSplitSize,
         array(ignoreBitfield) | ~array(filterBitfield_) | array(bitfield_) |
             array(useBitfield_),
         useBitfield_, blockLength_, blocks_);
   }
   else {
     return aria2::getInorderMissingUnusedIndex(
-        index, 0, blocks_, minSplitSize,
+        0, blocks_, minSplitSize,
         array(ignoreBitfield) | array(bitfield_) | array(useBitfield_),
         useBitfield_, blockLength_, blocks_);
   }
 }
 
-bool BitfieldMan::getInorderMissingUnusedIndex(
-    size_t& index, size_t startIndex, size_t endIndex, int32_t minSplitSize,
+std::optional<size_t> BitfieldMan::getInorderMissingUnusedIndex(
+    size_t startIndex, size_t endIndex, int32_t minSplitSize,
     const unsigned char* ignoreBitfield, size_t ignoreBitfieldLength) const
 {
   endIndex = std::min(endIndex, blocks_);
   if (filterEnabled_) {
     return aria2::getInorderMissingUnusedIndex(
-        index, startIndex, endIndex, minSplitSize,
+        startIndex, endIndex, minSplitSize,
         array(ignoreBitfield) | ~array(filterBitfield_) | array(bitfield_) |
             array(useBitfield_),
         useBitfield_, blockLength_, blocks_);
   }
   else {
     return aria2::getInorderMissingUnusedIndex(
-        index, startIndex, endIndex, minSplitSize,
+        startIndex, endIndex, minSplitSize,
         array(ignoreBitfield) | array(bitfield_) | array(useBitfield_),
         useBitfield_, blockLength_, blocks_);
   }
diff --git a/src/BitfieldMan.h b/src/BitfieldMan.h
index ffef7b0b..1937ddae 100644
--- a/src/BitfieldMan.h
+++ b/src/BitfieldMan.h
@@ -37,6 +37,7 @@
 
 #include "common.h"
 
+#include <optional>
 #include <vector>
 
 namespace aria2 {
@@ -107,7 +108,7 @@ public:
   bool hasMissingPiece(const unsigned char* bitfield, size_t len) const;
 
   // affected by filter
-  bool getFirstMissingUnusedIndex(size_t& index) const;
+  std::optional<size_t> getFirstMissingUnusedIndex() const;
 
   // Appends at most n missing unused index to out. This function
   // doesn't delete existing elements in out.  Returns the number of
@@ -120,7 +121,7 @@ public:
   // index is found. Otherwise returns false.
   //
   // affected by filter
-  bool getFirstMissingIndex(size_t& index) const;
+  std::optional<size_t> getFirstMissingIndex() const;
 
   // Stores missing bit index to index. index is selected so that it
   // divides longest missing bit subarray into 2 equally sized
@@ -128,9 +129,9 @@ public:
   // if such bit index is found. Otherwise returns false.
   //
   // affected by filter
-  bool getSparseMissingUnusedIndex(size_t& index, int32_t minSplitSize,
-                                   const unsigned char* ignoreBitfield,
-                                   size_t ignoreBitfieldLength) const;
+  std::optional<size_t> getSparseMissingUnusedIndex(
+      int32_t minSplitSize, const unsigned char* ignoreBitfield,
+      size_t ignoreBitfieldLength) const;
 
   // Stores missing bit index to index. This function first try to
   // select smallest index starting offsetIndex in the order:
@@ -146,10 +147,10 @@ public:
   // result.
   //
   // affected by filter
-  bool getGeomMissingUnusedIndex(size_t& index, int32_t minSplitSize,
-                                 const unsigned char* ignoreBitfield,
-                                 size_t ignoreBitfieldLength, double base,
-                                 size_t offsetIndex) const;
+  std::optional<size_t> getGeomMissingUnusedIndex(
+      int32_t minSplitSize, const unsigned char* ignoreBitfield,
+      size_t ignoreBitfieldLength, double base,
+      size_t offsetIndex) const;
 
   // Stores missing bit index to index. This function selects smallest
   // index of missing piece, considering minSplitSize.  Set bits in
@@ -157,19 +158,19 @@ public:
   // found. Otherwise returns false.
   //
   // affected by filter
-  bool getInorderMissingUnusedIndex(size_t& index, int32_t minSplitSize,
-                                    const unsigned char* ignoreBitfield,
-                                    size_t ignoreBitfieldLength) const;
+  std::optional<size_t> getInorderMissingUnusedIndex(
+      int32_t minSplitSize, const unsigned char* ignoreBitfield,
+      size_t ignoreBitfieldLength) const;
 
   // Just like getInorderMissingUnusedIndex() above, but limit the
   // search area in [startIndex, endIndex).  |endIndex| is normalized
   // to min(|endIndex|, blocks_)
   //
   // affected by filter
-  bool getInorderMissingUnusedIndex(size_t& index, size_t startIndex,
-                                    size_t endIndex, int32_t minSplitSize,
-                                    const unsigned char* ignoreBitfield,
-                                    size_t ignoreBitfieldLength) const;
+  std::optional<size_t> getInorderMissingUnusedIndex(
+      size_t startIndex, size_t endIndex, int32_t minSplitSize,
+      const unsigned char* ignoreBitfield,
+      size_t ignoreBitfieldLength) const;
 
   // affected by filter
   bool getAllMissingIndexes(unsigned char* misbitfield, size_t mislen) const;
